File: Basic_Linked_List.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node {
    int data;
    struct node* link;
} *NODE;

NODE createNode() {
    NODE newNode = (NODE)malloc(sizeof(struct node));
    if (newNode) {
        printf("Enter data for the new node: ");
        scanf("%d", &newNode->data);
        newNode->link = NULL;
    }
    return newNode;
}

void insertFront(NODE* head, NODE newNode) {
    if (newNode) {
        newNode->link = *head;
        *head = newNode;
    }
}

void insertRear(NODE* head, NODE newNode) {
    if (newNode) {
        if (*head == NULL) {
            *head = newNode;
        } else {
            NODE current = *head;
            while (current->link != NULL) {
                current = current->link;
            }
            current->link = newNode;
        }
    }
}

void deleteFront(NODE* head) {
    if (*head) {
        NODE temp = *head;
        *head = (*head)->link;
        free(temp);
    }
}

void deleteRear(NODE* head) {
    if (*head) {
        if ((*head)->link == NULL) {
            free(*head);
            *head = NULL;
        } else {
            NODE current = *head;
            while (current->link->link != NULL) {
                current = current->link;
            }
            free(current->link);
            current->link = NULL;
        }
    }
}

void printList(NODE head) {
    NODE current = head;
    printf("Linked List: ");
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->link;
    }
    printf("NULL\n");
}

int main() {
    NODE head = NULL;
    int choice;

    while (1) {
        printf("\nChoose an operation:\n");
        printf("1. Insert at the front\n");
        printf("2. Insert at the rear\n");
        printf("3. Delete from the front\n");
        printf("4. Delete from the rear\n");
        printf("5. Print the linked list\n");
        printf("6. Quit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                insertFront(&head, createNode());
                break;
            case 2:
                insertRear(&head, createNode());
                break;
            case 3:
                deleteFront(&head);
                break;
            case 4:
                deleteRear(&head);
                break;
            case 5:
                printList(head);
                break;
            case 6:
                exit(0);
            default:
                printf("Invalid choice. Please choose a valid option.\n");
        }
    }

    return 0;
}

File: Basic_Linked_List.py
class Node:
    def __init__(self, data):
        self.data = data
        self.link = None

def create_node():
    data = int(input("Enter data for the new node: "))
    return Node(data)

def insert_front(head, new_node):
    if new_node:
        new_node.link = head
        return new_node
    return head

def insert_rear(head, new_node):
    if new_node:
        if head is None:
            return new_node
        current = head
        while current.link:
            current = current.link
        current.link = new_node
    return head

def delete_front(head):
    if head:
        head = head.link
    return head

def delete_rear(head):
    if head:
        if not head.link:
            return None
        current = head
        while current.link.link:
            current = current.link
        current.link = None
    return head

def print_list(head):
    current = head
    print("Linked List: ", end='')
    while current:
        print(current.data, " -> ", end='')
        current = current.link
    print("NULL")

if __name__ == "__main__":
    head = None

    while True:
        print("\nChoose an operation:")
        print("1. Insert at the front")
        print("2. Insert at the rear")
        print("3. Delete from the front")
        print("4. Delete from the rear")
        print("5. Print the linked list")
        print("6. Quit")

        choice = int(input("Enter your choice: ")

        if choice == 1:
            head = insert_front(head, create_node())
        elif choice == 2:
            head = insert_rear(head, create_node())
        elif choice == 3:
            head = delete_front(head)
        elif choice == 4:
            head = delete_rear(head)
        elif choice == 5:
            print_list(head)
        elif choice == 6:
            break
        else:
            print("Invalid choice. Please choose a valid option.")
